File: backend/routes/quiz.py

from flask import Blueprint, request, jsonify
from models import User, QuizQuestion, UserQuiz, db
from routes.auth import token_required
from routes.tasks import check_for_new_badges
import json

quiz_bp = Blueprint('quiz', __name__)

# Get quiz questions
@quiz_bp.route('/quiz', methods=['GET'])
@token_required
def get_quiz_questions(current_user):
    # Get all quiz questions
    questions = QuizQuestion.query.all()
    
    # Format response
    questions_list = []
    for question in questions:
        questions_list.append({
            'id': question.id,
            'text': question.text,
            'options': json.loads(question.options)
        })
    
    return jsonify({
        'questions': questions_list
    }), 200

# Submit quiz answer
@quiz_bp.route('/quiz/submit', methods=['POST'])
@token_required
def submit_quiz_answer(current_user):
    data = request.get_json()
    
    # Check if required fields are provided
    if not all(k in data for k in ['question_id', 'selected_option']):
        return jsonify({'message': 'Missing required fields'}), 400
    
    question_id = data['question_id']
    selected_option = data['selected_option']
    
    # Check if question exists
    question = QuizQuestion.query.get(question_id)
    if not question:
        return jsonify({'message': 'Question not found'}), 404
    
    # Check if answer is correct
    is_correct = selected_option == question.correct_answer
    
    # Record quiz attempt
    quiz_attempt = UserQuiz(
        user_id=current_user.id,
        question_id=question_id,
        is_correct=is_correct
    )
    
    db.session.add(quiz_attempt)
    
    # Award points if answer is correct
    points_earned = 0
    if is_correct:
        points_earned = 15  # Award 15 points for correct answer
        current_user.points += points_earned
    
    db.session.commit()
    
    # Check for new badges
    new_badges = []
    if is_correct:
        new_badges = check_for_new_badges(current_user)
    
    return jsonify({
        'is_correct': is_correct,
        'correct_answer': question.correct_answer,
        'points_earned': points_earned,
        'total_points': current_user.points,
        'new_badges': new_badges
    }), 200

File: backend-nodejs/routes/quiz.js

const express = require('express');
const Question = require('../models/Question');
const User = require('../models/User');
const Badge = require('../models/Badge');
const auth = require('../middleware/auth');

const router = express.Router();

// Get quiz questions
router.get('/', auth, async (req, res) => {
  try {
    const questions = await Question.find({ isActive: true })
      .select('-correctAnswer')
      .limit(10);

    res.json({ questions });
  } catch (error) {
    console.error('Error fetching quiz questions:', error);
    res.status(500).json({ message: 'Server error fetching quiz questions' });
  }
});

// Submit quiz answer
router.post('/submit', auth, async (req, res) => {
  try {
    const { question_id, selected_option } = req.body;
    
    const question = await Question.findById(question_id);
    if (!question) {
      return res.status(404).json({ message: 'Question not found' });
    }

    const user = await User.findById(req.user._id);
    const isCorrect = question.correctAnswer === selected_option;
    
    // Record the answer
    user.quizAnswers.push({
      questionId: question_id,
      selectedOption: selected_option,
      isCorrect
    });

    let pointsEarned = 0;
    const newBadges = [];

    if (isCorrect) {
      pointsEarned = question.points;
      user.points += pointsEarned;

      // Check for new badges
      const badges = await Badge.find({ 
        pointsRequired: { $lte: user.points },
        isActive: true 
      });
      
      badges.forEach(badge => {
        const alreadyHasBadge = user.badges.some(
          ub => ub.badgeId.toString() === badge._id.toString()
        );
        
        if (!alreadyHasBadge) {
          user.badges.push({ badgeId: badge._id });
          newBadges.push(badge);
        }
      });
    }

    await user.save();

    res.json({
      is_correct: isCorrect,
      correct_answer: question.correctAnswer,
      points_earned: pointsEarned,
      total_points: user.points,
      new_badges: newBadges
    });
  } catch (error) {
    console.error('Error submitting quiz answer:', error);
    res.status(500).json({ message: 'Server error submitting answer' });
  }
});

module.exports = router;
